python/cudf/cudf/core/spillable_buffer.py

```diff
@@ -0,0 +1,445 @@
+    pointer (integer or void*) has been accessed outside of SpillableBuffer.
+    In this case, we cannot invalidate the device pointer by moving the data
+    to host.
+
+    A buffer can be exposed permanently at creation or by accessing the `.ptr`
+    property. To avoid this, one can use `.get_ptr()` instead, which support
+    exposing the buffer temporarily.
+
+    Parameters
+    ----------
+    data : buffer-like
+        An buffer-like object representing device or host memory.
+    exposed : bool, optional
+        Whether or not a raw pointer (integer or C pointer) has
+        been exposed to the outside world. If this is the case,
+        the buffer cannot be spilled.
+    manager : SpillManager
+        The manager overseeing this buffer.
+    """
+
+    _ptr_desc: Dict[str, Any]
+    _spill_locks: weakref.WeakSet
+
+    def __init__(
+        self,
+        data: Any,
+        exposed: bool,
+        manager: SpillManager,
+    ):
+        self._lock = RLock()
+        self._spill_locks = weakref.WeakSet()
+        self._exposed = exposed
+        self._last_accessed = time.monotonic()
+
+        # First, we extract the memory pointer, size, and owner.
+        # If it points to host memory we either:
+        #   - copy to device memory if exposed=True
+        #   - or create a new buffer that are marked as spilled already.
+        if isinstance(data, SpillableBuffer):
+            raise ValueError(
+                "Cannot create from a SpillableBuffer, "
+                "use __getitem__ to create a view"
+            )
+        if isinstance(data, rmm.DeviceBuffer):
+            self._ptr_desc = {"type": "gpu"}
+            self._ptr = data.ptr
+            self._size = data.size
+            self._owner = data
+        elif hasattr(data, "__cuda_array_interface__"):
+            self._ptr_desc = {"type": "gpu"}
+            ptr, size = get_ptr_and_size(data.__cuda_array_interface__)
+            self._ptr = ptr
+            self._size = size
+            self._owner = data
+        else:
+            if self._exposed:
+                self._ptr_desc = {"type": "gpu"}
+                ptr, size = get_ptr_and_size(
+                    numpy.array(data).__array_interface__
+                )
+                buf = rmm.DeviceBuffer(ptr=ptr, size=size)
+                self._ptr = buf.ptr
+                self._size = buf.size
+                self._owner = buf
+                # Since we are copying the data, we know that the device
+                # memory has not been exposed even if the original host
+                # memory has been exposed.
+                self._exposed = False
+            else:
+                data = memoryview(data)
+                if not data.c_contiguous:
+                    raise ValueError("memoryview must be C-contiguous")
+                self._ptr_desc = {"type": "cpu", "memoryview": data}
+                self._ptr = 0
```

Comment:
Currently, it is a `int` because of `Buffer`. We could make it `Optional[int]` and then let the property raise an exception if `None`.
```python
    @property
    def ptr(self) -> int:
        return self._ptr
```